### ISSUE TITLE
[Proposal] Automatic HSTS

### ISSUE THREAD

**Issue by Klooven** [caddyserver/caddy#4751]
It's again time for me to set up another Caddy server. Since I started using Caddy a few years ago my web server setup including setting up HTTPS has gotten much simpler and quicker (thanks Caddy!). While moving over my sites to HTTPS I've also started to set the `Strict-Transport-Security` (HSTS) header on these sites. Until this I've been using a simple snippet that I import for each site in my Caddyfile:

```
(hsts) {
  header Strict-Transport-Security max-age=63072000
}
```

This works great for a few sites, but when managing tens of sites in one Caddyfile, this can be a bit cumbersome (I've forgotten to set the header many times and managed to only include it on specific paths on some sites by mistake). As Caddy markets itself as the server with automatic HTTPS, I thought that automatic HSTS could be a great addition to the security measures that Caddy ships with by default.

## Why HSTS?

To quote [MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Strict-Transport-Security):

> If a website accepts a connection through HTTP and redirects to HTTPS, visitors may initially communicate with the non-encrypted version of the site before being redirected, if, for example, the visitor types http://www.foo.com/ or even just foo.com. This creates an opportunity for a man-in-the-middle attack. The redirect could be exploited to direct visitors to a malicious site instead of the secure version of the original site.
> 
> The HTTP Strict Transport Security header informs the browser that it should never load a site using HTTP and should automatically convert all attempts to access the site using HTTP to HTTPS requests instead.

In addition to being able to mitigate MITM attacks, which are mentioned above, HSTS can also (marginally) increase page load times and reduce strain on the server as clients won't try to access sites using HTTP after the first visit.

## Why automatic HSTS in Caddy?

- Caddy already serves content using HTTPS by default (when possible)
- Users of Caddy would profit from an extra security "benefit" without any extra config
- Users that already use HSTS don't have to manually set the `Strict-Transport-Security` header on each site

This would still be possible to disable (like the automatic HTTPS feature), e.g. through a global option:

```
{
  hsts off
}
```

## Examples of automatic HSTS "in the wild"

While I'm not aware of any (web) server providing an automatic HSTS feature, many common developer tools and services automatically set the `Strict-Transport-Security` header, e.g.:

- [Vercel](https://vercel.com/docs/concepts/edge-network/encryption#support-for-hsts) (static site hosting etc.)
- [nginx-proxy](https://github.com/nginx-proxy/nginx-proxy#how-ssl-support-works) (reverse proxy for Docker containers)

## An alternative opt-in method

The consequences of a possible automatic HSTS feature in Caddy probably need some discussion and thinking. If automatic HSTS would be deemed as a feature not worth shipping by default in Caddy for a reason or another, an opt-in method could also be considered. E.g. by giving users a global option to enable HSTS for all sites:

```
{
  hsts
}
```

This option could then be extended to support setting the `max-age` value and other supported values for the header. An opt-in method could also of course be implemented through a plugin for Caddy if there isn't enough demand to integrate it into Caddy's core.

---

I'd be happy to hear other people's view on this subject! Specifically whether an opt-out or opt-in version would make more sense, and in case of an opt-in version whether it should be integrated into Caddy or be made as a plugin.

**Comment by mholt** [comment]
Thanks for writing a thoughtful proposal!

I love the idea of HSTS -- and I would have had it be the default myself already -- if the consequences for getting it wrong (or predicting the future wrong) weren't so severe: your site gets bricked.

A less risky default is to have the _clients_ use HTTPS by default (try HTTPS connection first, then HTTP), and browsers are starting to do this. The latest Chrome and Firefox do this, I think, and even have opt-in HTTPS-only modes where HTTP results in an error.

Since we don't know the intent of the site owner regarding the future of their site, I don't think it's wise to enable HSTS by default in the server, implicitly. Imagine development sites, for example.

A global opt-in is _possibly_ doable, but I still think being explicit about the header is better. I'm not inclined to add a new global option for an HTTP feature that -- I hope -- will be deprecated in the coming years.

HSTS is also vulnerable to first-request manipulations, before the client first receives the HSTS header over HTTP, attackers can simply remove it. I think of HSTS as a band-aid over just using HTTPS first, much like I think of OCSP & cert revocation as a band-aid over short cert lifetimes.

**Comment by git001** [comment]
how about to add a further field like "hsts" into https://caddyserver.com/docs/json/apps/http/servers/automatic_https/ ( https://github.com/caddyserver/caddy/blob/master/modules/caddyhttp/autohttps.go ) as the hsts header should only be added when HTTPS is used and 

Suggestion for the field.
```
"hsts": "off"|"string"

# example
"hsts": "Strict-Transport-Security max-age=63072000"
```

@Klooven why not using https://caddyserver.com/docs/caddyfile/directives/import and only add the import to the server. Do you edit the Caddyfile manually?

**Comment by mholt** [comment]
@git001 That reminds me, that the implementation of auto-HSTS would possibly be even harder than auto-HTTPS. :sweat:

**Comment by Klooven** [comment]
@mholt Thanks for the response!

> I love the idea of HSTS -- and I would have had it be the default myself already -- **if the consequences for getting it wrong (or predicting the future wrong) weren't so severe: your site gets bricked**.

This is very true, especially with longer `max-age` values. On the contrary, using shorter durations as `max-age` would partly (or even mostly) defeat the purpose of HSTS.

If this feature _would ever_ become a part of Caddy, users should clearly be informed about the implications of the feature. But this would still increase the risk of mistakes when comparing to explicitly activating HSTS (setting the `Strict-Transport-Security` header).

> A less risky default is to have the clients use HTTPS by default (try HTTPS connection first, then HTTP), and browsers are starting to do this. **The latest Chrome and Firefox do this, I think, and even have opt-in HTTPS-only modes where HTTP results in an error.**

This was something new for me! Sounds like a great replacement for HSTS (at least in most cases)  -- without the risk of bricking your site. Without having looked at the specs for how the browsers do this I can come to think about at least one downside/lack of this "headerless HSTS" approach: as long as HTTPS isn't enforced in the browser (i.e. blocking HTTP), downgrade attacks could still be possible, because I assume the browser won't add a "HTTPS-only" flag automatically for a site that _has_ been accessible over HTTPS (this could brick the site in the same way as the STS header can).

On the other side, after receiving the `Strict-Transport-Security` header once, the browser will refuse to serve the domain through HTTP, mitigating possible downgrade attacks. (It is worth mentioning that attackers can't reset this by setting `max-age` to `0` on HTTP, as only STS headers sent through HTTPS are considered valid by browsers.)

As much as I would hope to see browsers being HTTPS-only for all sites, I'm not quite convinced that we'll see that in the few coming years (I'd be happy to predict wrong, though). And as I described above, I'm concerned about the "headerless HSTS'" ability to completely replace the `Strict-Transport-Security` header considering possible downgrade etc. attacks.

> A global opt-in is possibly doable, but I still think being explicit about the header is better. I'm not inclined to add a new global option for an HTTP feature that -- I hope -- will be deprecated in the coming years.

Referring to the comments above, I'm leaning more towards the direction of an opt-in version for this possible HSTS option. The question whether it should be integrated into Caddy's core boils down, in my opinion, to two questions:

- The relevancy of the `Strict-Transport-Security` header now and in the future
- The demand for the feature of Caddy's users

I cannot comment much on the second point, as I'm just one user out of many. When it comes to the relevancy of the STS header, though, I'm not fully convinced that it would be considered deprecated in just a few years, considering abovementioned possible attacks with "headerless HSTS" (aka. browsers connecting with HTTPS first).

That being said, if HSTS would be considered deprecated in the future, then it is so. There is no reason to keep deprecated features available to users for a prolonged time (Caddy has for example removed CLF logs from it's core), but I wouldn't refrain from implementing a feature just because it might be deprecated in the future.

---

@git001

> how about to add a further field like "hsts" into https://caddyserver.com/docs/json/apps/http/servers/automatic_https/ ( https://github.com/caddyserver/caddy/blob/master/modules/caddyhttp/autohttps.go ) as the hsts header should only be added when HTTPS is used

Sounds like a good way to possibly implement this into Caddy.

> why not using https://caddyserver.com/docs/caddyfile/directives/import and only add the import to the server. Do you edit the Caddyfile manually?

My current Caddyfile structure is like this:

```
(hsts) {
  header Strict-Transport-Security max-age=63072000
}

site1 {
  import hsts
  ...
}

site2 {
  import hsts
  ...
}

site3 {
  import hsts
  ...
}

...
```

I think this is currently the easiest way to implement HSTS for all sites. From what I've understood, Caddy doesn't provide a way to set "global headers", at least using the Caddyfile. Might be wrong on this?

---

**tl;dr** I'm just looking for a simple way for me and other Caddy users to enable HSTS, which is, in my opinion, still for now a relevant part in securing the web. If the implementation of this feature would, however, require huge refactoring of code and many hours of time, then it would probably be more wise to just set the STS headers manually using the `header` directive of Caddy on each site.

Thanks for all the comments!

**Comment by francislavoie** [comment]
Firefox: https://blog.mozilla.org/security/2021/08/10/firefox-91-introduces-https-by-default-in-private-browsing/

Chrome https://blog.chromium.org/2021/03/a-safer-default-for-navigation-https.html

**Comment by Klooven** [comment]
I respect your stance on the subject and appreciate that you took your time to read and comment on the proposal!

I _might_ look into the possibility to implement this as a plugin if I have enough time, but will for now continue setting the STS header manually. I'll of course also be waiting for the day when I can ditch the header all together 😁

**Comment by francislavoie** [comment]
Appreciate the thoughtful discussion!

I think we won't take action on this, so I'll close this.
